proposals/0000-type-vs-constraint.rst

```diff
@@ -0,0 +1,393 @@
+Generalising Type and Constraint
+==============================
+
+.. author:: Richard Eisenberg and Simon Peyton Jones
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/518>`_.
+.. sectnum::
+.. contents::
+
+Since version 8.0, GHC has supported a rich structure of inhabited types through the
+new primitive ``TYPE``, allowing
+unlifted types to be treated as first-class citizens, leading to the possibility
+of unlifted datatypes and newtypes, already available.
+
+This proposal extends this treatment to constraints, whose inhabitants
+can usefully be viewed as implicit parameters (typically type class dictionaries).
+Just like how the existence of ``TYPE`` directly opened the door to new innovations
+in datatype and newtype definitions, we expect the similar treatment of constraints
+to allow for strict classes and unlifted and unboxed implicit parameters.
+
+In addition, the new treatment proposed here allows GHC's internal language, Core,
+to more properly tell the difference between types and constraints, eliminating a
+troubling class of bugs that keeps recurring.
+
+Motivation
+----------
+
+1. With this proposal, we immediately allow unlifted and unboxed implicit parameters
```

Comment:
ImplicitParams seems to be widely viewed as a misfeature, or at least very niche. I'm not sure it would be accepted by the Steering Committee if it were proposed today.

With that in mind, I question how strong this motivation is. Is the lack of unlifted/unboxed ImplicitParams blocking real use cases?

Reply:
Well, it's pretty strong you like implicit parameters.  Ed Kmett wants this for sure.  I suppose someone could do some stats on Hackage to see how widely implicit parameters are used.   (Personally I don't see them as  a mis-feature, fwiw.)

Reply:
I do like ImplicitParams! But I've always felt I was in the minority in that opinion :)

Reply:
But not alone with it - for some things they are great :-)

Reply:
I based the design of my [`HUnit-approx`](https://hackage.haskell.org/package/HUnit-approx) library on implicit parameters. The feature shines there. And I thought the package was utterly unused, only to learn recently that a new-and-upcoming automatic differentiation library is using it. :) (I don't think that's public yet, so no link, sadly.)

Reply:
Implicit Parameters: "[Most of the following, still rather incomplete, documentation is due to Jeff Lewis.](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/implicit_parameters.html)" -- and the text there in the User Guide seems to be the same as for Hugs. So it's been "rather incomplete" for something like two decades.

@EKmett likes them, but is also [wary of gotchas](https://old.reddit.com/r/haskell/comments/m8o88a/who_still_uses_readert/grkh17q/) -- at least as of a year ago.

If IPs are a major motivation here, is it time to show them some love?